File: base/src/main/java/com/zhangteng/base/adapter/TreeRecyclerViewAdapter.java
package com.zhangteng.base.adapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.zhangteng.base.tree.Node;
import com.zhangteng.base.tree.TreeHelper;

import java.util.List;

/**
 * 树结构的列表适配器
 * Created by swing on 2018/6/29.
 */
public abstract class TreeRecyclerViewAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    protected Context mContext;
    /**
     * 存储所有可见的Node
     */
    protected List<Node> mNodes;
    protected LayoutInflater mInflater;
    /**
     * 存储所有的Node
     */
    protected List<Node> mAllNodes;

    /**
     * 点击的回调接口
     */
    private OnTreeNodeClickListener onTreeNodeClickListener;

    /**
     * @param mTree
     * @param context
     * @param datas
     * @param defaultExpandLevel 默认展开几级树
     * @throws IllegalArgumentException
     * @throws IllegalAccessException
     */
    public TreeRecyclerViewAdapter(RecyclerView mTree, Context context, List<T> datas, int defaultExpandLevel) throws IllegalArgumentException, IllegalAccessException {
        mContext = context;
        /**
         * 对所有的Node进行排序
         */
        mAllNodes = TreeHelper.getSortedNodes(datas, defaultExpandLevel);
        /**
         * 过滤出可见的Node
         */
        mNodes = TreeHelper.filterVisibleNode(mAllNodes);
        mInflater = LayoutInflater.from(context);

    }

    public void setOnTreeNodeClickListener(OnTreeNodeClickListener onTreeNodeClickListener) {
        this.onTreeNodeClickListener = onTreeNodeClickListener;
    }

    /**
     * 相应ListView的点击事件 展开或关闭某节点
     *
     * @param position
     */
    public void expandOrCollapse(int position) {
        Node n = mNodes.get(position);
        if (n != null) {// 排除传入参数错误异常
            if (!n.isLeaf()) {
                n.setExpand(!n.isExpand());
                mNodes = TreeHelper.filterVisibleNode(mAllNodes);
                notifyDataSetChanged();// 刷新视图
            }
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return getCreateViewHolder(parent, viewType);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, @SuppressLint("RecyclerView") final int position) {
        Node node = mNodes.get(position);
        getBindViewHolder(node, position, holder);
        // 设置内边距
        holder.itemView.setPadding(node.getLevel() * 40, 3, 3, 3);
        /**
         * 设置节点点击时，可以展开以及关闭；并且将ItemClick事件继续往外公布
         */
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                expandOrCollapse(position);
                if (onTreeNodeClickListener != null) {
                    onTreeNodeClickListener.onClick(mNodes.get(position), position);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return mNodes.size();
    }

    public abstract RecyclerView.ViewHolder getCreateViewHolder(ViewGroup parent, int viewType);

    public abstract void getBindViewHolder(Node node, int position, RecyclerView.ViewHolder holder);

    public interface OnTreeNodeClickListener {
        void onClick(Node node, int position);
    }
}

File: base/src/main/java/com/zhangteng/base/adapter/TreeRecyclerViewAdapter.kt
package com.zhangteng.base.adapter

import android.annotation.SuppressLint
import android.content.*
import com.zhangteng.base.base.BaseAdapter
import com.zhangteng.base.tree.*

/**
 * 树结构的列表适配器
 * Created by swing on 2018/6/29.
 */
abstract class TreeRecyclerViewAdapter<T>(data: MutableList<T?>?, defaultExpandLevel: Int = 1) :
    BaseAdapter<T, BaseAdapter.DefaultViewHolder>() {

    /**
     * 存储所有可见的Node
     */
    protected var mNodes: MutableList<Node<T?>?>?

    /**
     * 存储所有的Node
     */
    protected var mAllNodes: MutableList<Node<T?>?>?

    /**
     * 点击的回调接口
     */
    var onTreeNodeClickListener: OnTreeNodeClickListener? = null

    override fun onBindViewHolder(holder: DefaultViewHolder, item: T?, position: Int) {
        val node = mNodes?.get(position)
        onBindViewHolder(holder, node, position)

        /**
         * 设置节点点击时，可以展开以及关闭；并且将ItemClick事件继续往外公布
         */
        holder.itemView.setOnClickListener {
            expandOrCollapse(position)
            if (onTreeNodeClickListener != null) {
                onTreeNodeClickListener!!.onClick(mNodes?.get(position), position)
            }
        }
    }

    override fun getItemCount(): Int {
        return mNodes?.size ?: 0
    }

    abstract fun onBindViewHolder(
        holder: DefaultViewHolder?,
        node: Node<T?>?,
        position: Int
    )

    /**
     * 相应ListView的点击事件 展开或关闭某节点
     *
     * @param position
     */
    @SuppressLint("NotifyDataSetChanged")
    open fun expandOrCollapse(position: Int) {
        val n = mNodes?.get(position)
        if (n != null) { // 排除传入参数错误异常
            if (!n.isLeaf()) {
                n.setExpand(!n.isExpand())
                mNodes = TreeHelper.filterVisibleNode(mAllNodes)
                notifyDataSetChanged() // 刷新视图
            }
        }
    }

    /**
     * 获取显示的节点
     *
     * @param position
     */
    fun getVisibleNode(position: Int): Node<T?>? {
        return mNodes?.get(position)
    }

    interface OnTreeNodeClickListener {
        open fun <T> onClick(node: Node<T?>?, position: Int)
    }

    /**
     * @param data 原始数据
     * @param defaultExpandLevel 默认展开几级树
     */
    init {
        /**
         * 对所有的Node进行排序
         */
        mAllNodes = TreeHelper.getSortedNodes(data, defaultExpandLevel)
        /**
         * 过滤出可见的Node
         */
        mNodes = TreeHelper.filterVisibleNode(mAllNodes)
    }
}